### PR TITLE
refactor(lib): `RemoteMethods` follow the dispatch pattern

### LINKED ISSUE #1744: Questions around `/push` endpoints
Right now, we have `/push` do 3 things depending on the http method

1) GET - returns a list of your datasets on the given remote as `[]VersionInfo`
2) POST - expected behavior, pushes a dataset to a remote
3) DELETE - removes a dataset from a remote
Seems like we need to add a few more endpoints to our spec to cover cases 1 & 3 or at least make sure they are covered under other endpoints.

Should we add `/remote/list` and `/remote/remove`, or should these be folded into `/list` and `/ds/remove`?

### PR DESCRIPTION
**Pull request by ramfox**
- [x] add `Name` & `Attributes` methods on `RemoteMethods`
- [x] add `Remote` method to Instance
- [x] add `remoteImpl`
- [x] add `Preview`, `Feeds`, `Push`, `Pull` to `remoteImpl` that take a scope
- [x] dispatch-ify `Preview`, `Feeds`, `Push`, `Pull` methods
- [x] add not-yet-implemented `List` method
- [x] register `Remote` with dipatch
- [x] replace `NewRemoteMethods` with `inst.Remote`
- [x] replace inst in `RemoteMethods` with `d dispatcher`
- [x] add `AERemoteList` => `/remote/list` & `AERemoteDelete` => `/remote/delete` endpoints to lib/api.go
- [x] use `NewHTTPRequestMethods` for all api handlers

--- 

turns out the remote listing method was not actually getting a list of datasets from a remote, but rather listing the datasets in your repo that are public. Since this is not the functionality we expect or want, we are removing the concept of listing from a remote until it comes up in practice & can be written with attention.

Closes #1744

### CODE REVIEW

File: lib/api.go [hunk 1]
```diff
@@ -115,6 +115,10 @@ const (
 	AEFeeds = APIEndpoint("/feeds")
 	// AEPreview fetches a dataset preview from the registry
 	AEPreview = APIEndpoint("/preview")
+	// AERemoteList fetchs a list of the users dataset published on a given remote
+	AERemoteList = APIEndpoint("/remote/list")
+	// AERemoteRemove removes a dataset from a given remote
+	AERemoteRemove = APIEndpoint("/remote/remote")
```
Comment (Arqu) [inline lib/api.go:121]: `/remote/remote` should be `/remote/remove`

File: lib/remote.go [hunk 2]
```diff
@@ -37,26 +43,12 @@ type FeedsParams struct {
...
-func (r *RemoteMethods) Feeds(ctx context.Context, p *FeedsParams) (map[string][]dsref.VersionInfo, error) {
-	if r.inst.http != nil {
-		res := map[string][]dsref.VersionInfo{}
-		err := r.inst.http.Call(ctx, AEFeeds, p, &res)
-		if err != nil {
-			return nil, err
-		}
-		return res, nil
+func (m RemoteMethods) Feeds(ctx context.Context, p *FeedsParams) (map[string][]dsref.VersionInfo, error) {
+	got, _, err := r.d.Dispatch(ctx, dispatchMethodName(r, "feeds"), p)
```
Comment (Arqu) [inline lib/remote.go:47]: `r.d.Dispatch(...` does not work should be `m.d.Dispatch(...` if I got this right.

File: lib/remote.go [hunk 3]
```diff
@@ -84,92 +76,115 @@ func (p *PreviewParams) UnmarshalFromRequest(r *http.Request) error {
...
-func (r *RemoteMethods) Preview(ctx context.Context, p *PreviewParams) (*dataset.Dataset, error) {
-	if r.inst.http != nil {
-		res := &dataset.Dataset{}
-		err := r.inst.http.Call(ctx, AEPreview, p, res)
-		if err != nil {
-			return nil, err
-		}
-		return res, nil
+func (m RemoteMethods) Preview(ctx context.Context, p *PreviewParams) (*dataset.Dataset, error) {
+	got, _, err := r.d.Dispatch(ctx, dispatchMethodName(r, "preview"), p)
```
Comment (Arqu) [inline lib/remote.go:80]: `r` > `m`

File: lib/remote.go [hunk 4]
```diff
@@ -84,92 +76,115 @@ func (p *PreviewParams) UnmarshalFromRequest(r *http.Request) error {
...
+	Ref    string `schema:"refstr" json:"refstr"`
+	Remote string
+	// All indicates all versions of a dataset and the dataset namespace should
+	// be either published or removed
+	All bool
+}
+
+// Push posts a dataset version to a remote
+func (m RemoteMethods) Push(ctx context.Context, p *PushParams) (*dsref.Ref, error) {
+	got, _, err := r.d.Dispatch(ctx, dispatchMethodName(r, "push"), p)
```
Comment (Arqu) [inline lib/remote.go:98]: `r` > `m`

File: lib/remote.go [hunk 5]
```diff
@@ -84,92 +76,115 @@ func (p *PreviewParams) UnmarshalFromRequest(r *http.Request) error {
...
+	if res, ok := got.(*dsref.Ref); ok {
+		return res, err
+	}
+	return nil, dispatchReturnError(got, err)
+}
+
+// Remove asks a remote to remove a dataset
+func (m RemoteMethods) Remove(ctx context.Context, p *PushParams) (*dsref.Ref, error) {
+	got, _, err := r.d.Dispatch(ctx, dispatchMethodName(r, "remove"), p)
```
Comment (Arqu) [inline lib/remote.go:107]: `r` > `m`

File: lib/remote.go [hunk 6]
```diff
@@ -84,92 +76,115 @@ func (p *PreviewParams) UnmarshalFromRequest(r *http.Request) error {
...
+	if res, ok := got.(*dsref.Ref); ok {
+		return res, err
 	}
+	return nil, dispatchReturnError(got, err)
+}
 
+// List returns a list of datasets this user has pushed to the remote
+func (m RemoteMethods) List(ctx context.Context, p *ListParams) ([]dsref.VersionInfo, error) {
+	got, _, err := r.d.Dispatch(ctx, dispatchMethodName(r, "list"), p)
```
Comment (Arqu) [inline lib/remote.go:116]: `r` > `m`

**Review by Arqu — CHANGES_REQUESTED** [review]
👍 looking forward to rebase the refactoring on top of this :)

File: lib/remote.go (hunk 2 above)
Comment (ramfox) [reply lib/remote.go:47]: yea, i squashed a commit and i guess did it wrong, and it lost this round of work :/

File: lib/api.go (hunk 1 above)
Comment (ramfox) [reply lib/api.go]: THANK YOU

**Review by Arqu — APPROVED** [review]
🚀

**Pull request merged**